src/node_osrm.cpp

```diff
@@ -375,257 +756,220 @@ struct RunQueryBaton
-    auto params = argumentsToParameter(args, false);
+    table_parameters_ptr params = argumentsToTableParameter(args, true);
     if (!params)
         return;
 
-    v8::Local<v8::Object> obj = Nan::To<v8::Object>(args[0]).ToLocalChecked();
+    BOOST_ASSERT(params->IsValid());
 
-    v8::Local<v8::Value> sources = obj->Get(Nan::New("sources").ToLocalChecked());
-    if (!sources->IsArray() && !sources->IsUndefined())
-    {
-        Nan::ThrowError("sources must be an array of coordinates (or undefined)");
-        return;
-    }
-    v8::Local<v8::Value> destinations = obj->Get(Nan::New("destinations").ToLocalChecked());
-    if (!destinations->IsArray() && !destinations->IsUndefined())
-    {
-        Nan::ThrowError("destinations must be an array of coordinates (or undefined)");
-        return;
-    }
+    Run<TableQueryBaton>(args, std::move(params), AsyncRunTable, AfterRunTable);
+}
 
-    if (destinations->IsUndefined() != sources->IsUndefined())
-    {
-        Nan::ThrowError("Both sources and destinations need to be specified");
+void Engine::nearest(const Nan::FunctionCallbackInfo<v8::Value> &args)
+{
+    Nan::HandleScope scope;
+    nearest_parameters_ptr params = argumentsToNearestParameter(args, false);
+    if (!params)
         return;
-    }
 
-    if (!destinations->IsUndefined() && !sources->IsUndefined() && params->coordinates.size() > 0)
+    BOOST_ASSERT(params->IsValid());
+
+    Run<NearestQueryBaton>(args, std::move(params), AsyncRunNearest, AfterRunNearest);
+}
+
+template <typename BatonType, typename ParamType, typename AsyncType, typename AfterType>
+void Engine::Run(const Nan::FunctionCallbackInfo<v8::Value> &args, ParamType params, AsyncType asnyc_fn, AfterType after_fn)
+{
+    Nan::HandleScope scope;
+    v8::Local<v8::Value> callback = args[args.Length() - 1];
+
+    if (!callback->IsFunction())
     {
-        Nan::ThrowError("You can either specify sources and destinations, or coordinates");
+        Nan::ThrowTypeError("last argument must be a callback function");
         return;
     }
 
-    if (sources->IsArray())
+    auto closure = new BatonType();
+    closure->request.data = closure;
+    closure->machine = ObjectWrap::Unwrap<Engine>(args.This());
+    closure->params = std::move(params);
+    closure->error = "";
+    closure->cb.Reset(callback.As<v8::Function>());
+    uv_queue_work(uv_default_loop(), &closure->request, asnyc_fn,
+                  reinterpret_cast<uv_after_work_cb>(after_fn));
+    closure->machine->Ref();
+    return;
+}
+
+void Engine::ParseResult(const osrm::engine::Status result_status_code, osrm::json::Object result)
+{
+    const auto message_iter = result.values.find("status_message");
+    const auto end_iter = result.values.end();
+    const auto result_code = static_cast<int>(result_status_code);
+
+    // 4xx : Invalid request
+    // 207 : No route found
+    // 208 : No edge found
+    if (result_code / 100 == 4 || result_code == 207 || result_code == 208)
```

Comment:
207 and 208 don't exist anymore, you can probably simplify here to a `== 400` check.